core/trino-main/src/test/java/io/trino/server/security/oauth2/TestOAuth2WebUiAuthenticationFilterWithUserinfo.java

```diff
@@ -0,0 +1,411 @@
+import okhttp3.CookieJar;
+import okhttp3.HttpUrl;
+import okhttp3.JavaNetCookieJar;
+import okhttp3.OkHttpClient;
+import okhttp3.Request;
+import okhttp3.Response;
+import org.testng.annotations.AfterClass;
+import org.testng.annotations.BeforeClass;
+import org.testng.annotations.Test;
+
+import java.io.IOException;
+import java.net.CookieManager;
+import java.net.CookieStore;
+import java.net.HttpCookie;
+import java.net.MalformedURLException;
+import java.net.URI;
+import java.net.URL;
+import java.security.KeyPairGenerator;
+import java.security.NoSuchAlgorithmException;
+import java.time.Duration;
+import java.time.Instant;
+import java.util.List;
+import java.util.Optional;
+import java.util.UUID;
+
+import static io.trino.client.OkHttpUtil.setupInsecureSsl;
+import static io.trino.server.security.oauth2.TokenEndpointAuthMethod.CLIENT_SECRET_BASIC;
+import static io.trino.server.ui.OAuthWebUiCookie.OAUTH2_COOKIE;
+import static javax.servlet.http.HttpServletResponse.SC_OK;
+import static javax.ws.rs.core.HttpHeaders.LOCATION;
+import static javax.ws.rs.core.Response.Status.OK;
+import static javax.ws.rs.core.Response.Status.SEE_OTHER;
+import static javax.ws.rs.core.Response.Status.UNAUTHORIZED;
+import static org.assertj.core.api.Assertions.assertThat;
+import static org.testng.Assert.assertEquals;
+
+public class TestOAuth2WebUiAuthenticationFilterWithUserinfo
+{
+    private static final Duration TTL_ACCESS_TOKEN_IN_SECONDS = Duration.ofSeconds(10);
+
+    private static final String TRINO_CLIENT_ID = "trino-client";
+    private static final String TRINO_CLIENT_SECRET = "trino-secret";
+    private static final String TRINO_AUDIENCE = TRINO_CLIENT_ID;
+    private static final String ADDITIONAL_AUDIENCE = "https://external-service.com";
+    private static final String TRUSTED_CLIENT_ID = "trusted-client";
+    private static final String TRUSTED_CLIENT_SECRET = "trusted-secret";
+    private static final String UNTRUSTED_CLIENT_ID = "untrusted-client";
+    private static final String UNTRUSTED_CLIENT_SECRET = "untrusted-secret";
+    private static final String UNTRUSTED_CLIENT_AUDIENCE = "https://untrusted.com";
+
+    private final Logging logging = Logging.initialize();
+    private final OkHttpClient httpClient;
+    private TestingHydraIdentityProvider hydraIdP;
+
+    private TestingTrinoServer server;
+    private URI serverUri;
+    private URI uiUri;
+
+    public TestOAuth2WebUiAuthenticationFilterWithUserinfo()
+    {
+        OkHttpClient.Builder httpClientBuilder = new OkHttpClient.Builder();
+        setupInsecureSsl(httpClientBuilder);
+        httpClientBuilder.followRedirects(false);
+        httpClient = httpClientBuilder.build();
+    }
+
+    @BeforeClass
+    public void setup()
+            throws Exception
+    {
+        logging.setLevel(TestOAuth2WebUiAuthenticationFilterWithUserinfo.class.getName(), Level.DEBUG);
+
+        hydraIdP = new TestingHydraIdentityProvider(TTL_ACCESS_TOKEN_IN_SECONDS);
```

Comment:
Is the access token really opaque? It looks it's using JWT token still. Take a look at: 
`.withEnv("STRATEGIES_ACCESS_TOKEN", "jwt")` in `TestingHydraIdentityProvider`

Reply:
The token actually being opaque is irrelevant, a JWT access token is still an access token which the `/userinfo` endpoint must accept and return a proper response for. That said this test has been modified to have Hydra use opaque tokens.